fish-rust/src/abbrs.rs

```diff
@@ -0,0 +1,479 @@
+#![allow(clippy::extra_unused_lifetimes, clippy::needless_lifetimes)]
+use std::{
+    collections::HashSet,
+    sync::{Arc, Mutex, MutexGuard},
+};
+
+use crate::{
+    wchar::L,
+    wchar_ffi::{WCharFromFFI, WCharToFFI},
+};
+use cxx::{CxxWString, UniquePtr};
+use once_cell::sync::Lazy;
+pub use widestring::{Utf32Str as wstr, Utf32String as WString};
+
+use crate::abbrs::abbr_ffi::abbrs_replacer_t;
+use crate::ffi::re::regex_t;
+use crate::parse_constants::SourceRange;
+
+use self::abbr_ffi::{abbreviation_t, abbrs_position_t, abbrs_replacement_t};
+
+#[cxx::bridge]
+mod abbr_ffi {
+    extern "C++" {
+        include!("re.h");
+        include!("parse_constants.h");
+
+        type SourceRange = crate::parse_constants::SourceRange;
+    }
+
+    enum abbrs_position_t {
+        command,
+        anywhere,
+    }
+
+    struct abbrs_replacer_t {
+        replacement: UniquePtr<CxxWString>,
+        is_function: bool,
+        set_cursor_marker: UniquePtr<CxxWString>,
+        has_cursor_marker: bool,
+    }
+
+    struct abbrs_replacement_t {
+        range: SourceRange,
+        text: UniquePtr<CxxWString>,
+        cursor: usize,
+        has_cursor: bool,
+    }
+
+    struct abbreviation_t {
+        key: UniquePtr<CxxWString>,
+        replacement: UniquePtr<CxxWString>,
+        is_regex: bool,
+    }
+
+    extern "Rust" {
+        type GlobalAbbrs<'a>;
+
+        #[cxx_name = "abbrs_list"]
+        fn abbrs_list_ffi() -> Vec<abbreviation_t>;
+
+        #[cxx_name = "abbrs_match"]
+        fn abbrs_match_ffi(token: &CxxWString, position: abbrs_position_t)
+            -> Vec<abbrs_replacer_t>;
+
+        #[cxx_name = "abbrs_has_match"]
+        fn abbrs_has_match_ffi(token: &CxxWString, position: abbrs_position_t) -> bool;
+
+        #[cxx_name = "abbrs_replacement_from"]
+        fn abbrs_replacement_from_ffi(
+            range: SourceRange,
+            text: &CxxWString,
+            set_cursor_marker: &CxxWString,
+            has_cursor_marker: bool,
+        ) -> abbrs_replacement_t;
+
+        #[cxx_name = "abbrs_get_set"]
+        unsafe fn abbrs_get_set_ffi<'a>() -> Box<GlobalAbbrs<'a>>;
+        unsafe fn add<'a>(
+            self: &mut GlobalAbbrs<'_>,
+            name: &CxxWString,
+            key: &CxxWString,
+            replacement: &CxxWString,
+            position: abbrs_position_t,
+            from_universal: bool,
+        );
+        unsafe fn erase<'a>(self: &mut GlobalAbbrs<'_>, name: &CxxWString);
+    }
+}
+
+static abbrs: Lazy<Arc<Mutex<AbbreviationSet>>> =
+    Lazy::new(|| Arc::new(Mutex::new(Default::default())));
+
+pub fn with_abbrs<R>(cb: impl FnOnce(&AbbreviationSet) -> R) -> R {
+    let abbrs_g = abbrs.lock().unwrap();
+    cb(&abbrs_g)
+}
+
+pub fn with_abbrs_mut<R>(cb: impl FnOnce(&mut AbbreviationSet) -> R) -> R {
+    let mut abbrs_g = abbrs.lock().unwrap();
+    cb(&mut abbrs_g)
+}
```

Comment:
Let me know if I should keep this or Use `abbrs_get_set` everywhere in rust also

Reply:
with_abbrs seems fine, I don't see any particular reason for or against it